asserts/asserts.go

```diff
@@ -257,10 +277,114 @@ func buildAssertion(headers map[string]string, body, content, signature []byte)
 	return assert, nil
 }
 
+func writeHeader(buf *bytes.Buffer, headers map[string]string, name string) {
+	buf.WriteString(name)
+	buf.WriteString(": ")
+	buf.WriteString(headers[name])
+	buf.WriteByte('\n')
+}
+
+var marker = struct{}{}
+
+// TODO: expose this in some form on Database appropriately
+func buildAndSign(assertType AssertionType, headers map[string]string, body []byte, privKey *packet.PrivateKey) (Assertion, error) {
+	finalHeaders := make(map[string]string, len(headers))
+	for name, value := range headers {
+		finalHeaders[name] = value
+	}
+	bodyLength := len(body)
+	finalBody := make([]byte, bodyLength)
+	copy(finalBody, body)
+	finalHeaders["type"] = string(assertType)
+	finalHeaders["body-length"] = strconv.Itoa(bodyLength)
+
+	if _, err := checkMandatory(finalHeaders, "authority-id"); err != nil {
+		return nil, err
+	}
+
+	reg, err := checkAssertType(assertType)
+	if err != nil {
+		return nil, err
+	}
+
+	revision, err := checkRevision(finalHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	buf := new(bytes.Buffer)
+	writeHeader(buf, finalHeaders, "type")
+	writeHeader(buf, finalHeaders, "authority-id")
+	if revision > 0 {
+		writeHeader(buf, finalHeaders, "revision")
+	} else {
+		delete(finalHeaders, "revision")
+	}
+	written := map[string]struct{}{
+		"type":         marker,
+		"authority-id": marker,
+		"revision":     marker,
+		"body-length":  marker,
+	}
+	for _, primKey := range reg.primaryKey {
+		if _, err := checkMandatory(finalHeaders, primKey); err != nil {
+			return nil, err
+		}
+		writeHeader(buf, finalHeaders, primKey)
+		written[primKey] = marker
+	}
+
+	// emit other headers in lexicographic order
+	otherKeys := make([]string, 0, len(finalHeaders))
+	for name := range finalHeaders {
+		_, ok := written[name]
+		if !ok {
+			otherKeys = append(otherKeys, name)
+		}
+	}
+	sort.Strings(otherKeys)
+	for _, k := range otherKeys {
+		writeHeader(buf, finalHeaders, k)
+	}
+
+	// body-length and body
+	if bodyLength > 0 {
+		writeHeader(buf, finalHeaders, "body-length")
+	} else {
+		delete(finalHeaders, "body-length")
+	}
+	if bodyLength > 0 {
+		buf.Grow(bodyLength + 2)
+		buf.WriteString("\n\n")
+		buf.Write(finalBody)
```

Comment:
There's no need to copy body into finalBody, since this is the only time we look at it. So, no inconsistency is possible in terms of the underlying array being mutable. Or at least, there's no further damage possible, in comparison to doing the copy once above.